### PR TITLE
ci: set up lighthouse

### PR DESCRIPTION
**Pull request by Splaktar**
ci: set up lighthouse

- define CircleCI version
- update NodeJS docker image
- use the branch as part of the cache_key
- add the circleci/build-tools orb
- store the build output for reuse in other jobs
- add new job to run the lighthouse a11y audits on various pages
- add new job to run all of the lighthouse audits on the home page
- add typing and comments to the original scripts
- update to the latest firebase-tools

Screen Shot 2021-01-12 at 15 38 22

Screen Shot 2021-01-12 at 15 39 22

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -19,19 +31,41 @@ var_5: &yarn_install
...
 
-# Job step for checking out the source code from GitHub. This also ensures that the source code
-# is rebased on top of master.
-var_6: &checkout_code
-  checkout:
-    # After checkout, rebase on top of master. By default, PRs are not rebased on top of master,
-    # which we want. See https://discuss.circleci.com/t/1662
-    post: git pull --ff-only origin "refs/pull/${CI_PULL_REQUEST//*pull\//}/merge"
+# https://circleci.com/docs/2.0/workflows/#using-workspaces-to-share-data-among-jobs
+# https://circleci.com/blog/deep-diving-into-circleci-workspaces/
```
Comment (josephperrott) [inline .circleci/config.yml:35]: Should we rename this to `var_4`

File: tools/audit-web-app.js [hunk 2]
```diff
@@ -0,0 +1,200 @@
+#!/bin/env node
+'use strict';
+
+/**
+ * Originally from https://github.com/angular/angular/blob/ab8199f7c909eaa8937d293ab44405fe263417cd/aio/scripts/audit-web-app.js
+ *
+ * Usage:
+ * ```sh
+ * node scripts/audit-web-app <url> <min-scores> [<log-file>]
```
Comment (josephperrott) [inline tools/audit-web-app.js:9]: the usage should be `node tools/audit-web-app...`

File: tools/audit-web-app.js [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        [ {passName : 'defaultPass', pauseAfterLoadMs : WAIT_FOR_SW_DELAY} ],
+  };
+
+  console.log(`Running web-app audits for '${url}'...`);
+  console.log(`  Audit categories: ${lhFlags.onlyCategories.join(', ')}`);
+
+  // If testing on HTTP, skip HTTPS-specific tests.
+  // (Note: Browsers special-case localhost and run ServiceWorker even on HTTP.)
+  if (isOnHttp)
+    skipHttpsAudits(lhConfig);
```
Comment (josephperrott) [inline tools/audit-web-app.js:78]:
```suggestion
  if (isOnHttp) {
    skipHttpsAudits(lhConfig);
  }
```

File: tools/audit-web-app.js [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+  logger.setLevel(lhFlags.logLevel);
+
+  try {
+    console.log('');
```
Comment (josephperrott) [inline tools/audit-web-app.js:83]: Is this necessary?  At minimum if we want a blank line, we can just have `console.log()`

File: tools/audit-web-app.js [hunk 5]
```diff
@@ -0,0 +1,200 @@
+#!/bin/env node
```
Comment (josephperrott) [inline tools/audit-web-app.js:1]: Can you add comments to this file to express what functions are doing and variables are representing?

File: tools/test-docs-a11y.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
+#!/bin/env node
+'use strict';
+
+/**
+ * Originally from https://github.com/angular/angular/blob/b48eabddb83900496e5e00fb98e192c6c49dff26/aio/scripts/test-aio-a11y.js
+ *
+ * Usage:
+ * ```sh
+ * node scripts/test-docs-a11y <origin>
```
Comment (josephperrott) [inline tools/test-docs-a11y.js:9]: `node tools/test-docs-a11y.js <origin>`

**Review by josephperrott — CHANGES_REQUESTED** [review]

**Comment by Splaktar** [comment]
@josephperrott thank you for the good feedback! It's all been addressed.

**Review by josephperrott — APPROVED** [review]
LGTM

**Pull request merged**